framework/decode/vulkan_ascii_consumer_base.h

```diff
@@ -109,22 +111,52 @@ class VulkanAsciiConsumerBase : public VulkanConsumer
 
   protected:
     template <typename ToStringFunctionType>
-    inline void WriteApiCallToFile(const ApiCallInfo&   call_info,
-                                   const std::string&   functionName,
-                                   util::ToStringFlags  toStringFlags,
-                                   uint32_t&            tabCount,
-                                   uint32_t             tabSize,
-                                   ToStringFunctionType toStringFunction)
+    inline void WriteApiCallToFile(const ApiCallInfo& call_info,
+                                   const std::string& functionName,
+                                   /// @todo Remove toStringFlags, tabCount, tabSize
+                                   util::ToStringFlags        toStringFlags,
+                                   const uint32_t&            tabCount,
+                                   const uint32_t             tabSize,
+                                   const ToStringFunctionType toStringFunction,
+                                   /// The formatted return string value excluding trailing comma
+                                   const char* const return_val = nullptr)
```

Comment:
This feels like an inconsistency against the other `std::string`s being passed around.  Can this be a `std::string&` and just check for `.empty()` instead of a `nullptr`?

Reply:
It totally could be a string ref. This is a deliberate choice. We have so many temporary strings being passed around already. If I make it a string ref, every caller for a function that doesn't have a return value, would be creating and passing in another temporary (albeit empty) string.

By making a first step away from string objects here it also opens the possibility of refactoring the callers not to use string objects. For instance, in this case, callers are converting VkResult enums into their text representation through `template <> std::string ToString<VkResult>(const VkResult& value, ToStringFlags, uint32_t, uint32_t)`, a non-inline function, which is a big switch causing `const char * const` strings to be copied into string objects\* only to be returned and copied around on the stack before being copied back out of them into a file buffer. Most of the `VkResult` strings (although not the most common ones) are too long for the short string optimisation to apply so we are talking about a gratuitous heap alloc and immediate free too. But those C-strings could have had their addresses passed around in registers before being copied directly into the C file buffers. There never needed to have been a string object involved.

If anything, I'd change the functionName argument to not be a `std::string&` either. That is a heap alloc, at least one string copy, and a heap free just to to pass the some data that started off static const one level into a function where it is fprintfed into the buffer of a file and then discarded, with no longer lifetime management issues to warrant the use of a RAII wrapper like `std::string`.

If you say definitely make it `std::string&`, it will be a `std::string&`, but I do need to double check this :sweat_smile: .

\*
```cpp
switch (value) {
    case VK_SUCCESS: return "VK_SUCCESS";
    case VK_NOT_READY: return "VK_NOT_READY";
    case VK_TIMEOUT: return "VK_TIMEOUT";
    case VK_EVENT_SET: return "VK_EVENT_SET";
    case VK_EVENT_RESET: return "VK_EVENT_RESET";
   ...
```

Reply:
I will try to convince you.  :)  I've spent a little time with Compiler Explorer and it looks like passing `std::string()` or `""` to a `std::string&` doesn't make a heap allocation in MSVC, Clang, or GCC on x64; nor in Clang on my M1 arm Mac.  (https://godbolt.org/z/dv4nb48zj - try stdin as "0" and "foobar" on separate lines, or "1" and "foobar", or "1" and "foobarfoobarfoobar".)  CE doesn't seem to have any working arm64 executors, but I will bet that the behavior is similar on Android.

The temporary for `string()` or `string` constructed with `""` is likely allocated on the stack, and then with any inlining `.empty()` in a called function is probably a comparison of the string length with 0 or possibly entirely skipped.  These may have few more instructions than passing and comparing with `nullptr` but I think these are insignificant compared to the amount of `string` and `stringstream` copying going on.

However, the non-empty return value strings _are_ allocated in this PR already, so passing a reference to them isn't going to change that, and they probably need to remain so because they'll be formatted.  Interestingly, small strings (looks like 15 characters or less) appear to not be heap allocated, probably stored in a small buffer in the string itself, but I haven't dug far enough to know.  You can test this with the Compiler Explorer snippet too, if you'd like!

My view is that receiving a `const string&` here is going to look more consistent, will be potentially safer, and isn't less efficient in any real way.  I welcome a counterpoint!

Reply:
> I will try to convince you.  :)  I've spent a little time with Compiler Explorer and it looks like passing `std::string()` or `""` to a `std::string&` doesn't make a heap allocation in MSVC, Clang, or GCC on x64; nor in Clang on my M1 arm Mac.  (https://godbolt.org/z/dv4nb48zj - try stdin as "0" and "foobar" on separate lines, or "1" and "foobar", or "1" and "foobarfoobarfoobar".)  CE doesn't seem to have any working arm64 executors, but I will bet that the behavior is similar on Android.

True, hence my "(albeit empty)" remark.

> The temporary for `string()` or `string` constructed with `""` is likely allocated on the stack, and then with any inlining `.empty()` in a called function is probably a comparison of the string length with 0 or possibly entirely skipped.  These may have few more instructions than passing and comparing with `nullptr` but I think these are insignificant compared to the amount of `string` and `stringstream` copying going on.

An empty string is still an instance of a class that is around 32 bytes on typical implementations (ptr,size, 15 bytes of short string buffer and space for a zero).
This is as you say living on the stack.
It is definitely more work to deal with such an object than with a plain-old-data register-sized pointer.
For efficiency we are relying on total elision of the strings' existance, which is possible with inlining.
However, this function is only inlinable because it is a template, which I don't think it ultimately should be after more refactoring to avoid code bloat \*.
I shouldn't have brought up the empty strings in passing though as they are not a big deal.

> However, the non-empty return value strings _are_ allocated in this PR already, so passing a reference to them isn't going to change that, and they probably need to remain so because they'll be formatted.  Interestingly, small strings (looks like 15 characters or less) appear to not be heap allocated, probably stored in a small buffer in the string itself, but I haven't dug far enough to know.  You can test this with the Compiler Explorer snippet too, if you'd like!

I don't think the return enum strings need to remain `std::string`s. They take part in formating by being the read-only source of a copy operation into a larger string.
If this function takes a `std::string` ref, there would be less point changing upstream code to not use them as it would only end up creating a `std::string` for this consumer function at the last moment.
With this interface, the caller can just do a `.c_str()` if it really wants to use `std::string` upstream of here, since that's what this function has to do internally anyway if it takes a `std::string` instead of a `const char*`.

Agreed that every compiler we care about implements the small/short string optimisation.


> My view is that receiving a `const string&` here is going to look more consistent, will be potentially safer, and isn't less efficient in any real way.  I welcome a counterpoint!

* Consistent: I want to change the other `std::string` to be a `const char*`, which will also be consistent :).
* Safety: We have strings sitting in the read-only data section of our executable, loaded into read-only data at startup, never modified, never deleted. The only thing we need to do is get them copied into a file buffer at the right moment, with the least ceremony possible. There are no lifetime issues.
* Efficiency: Right here in this PR, maybe efficiency isn't helped much, but it opens the door to callers not using `std::string` either. Eliminating temp `std::strings` as much as possible (and no more) from where we are now is like introducing `const` into a codebase that hasn't been using it. You need to find the leaves and work back from there. This is one of those leaves.

Stepping back, I'm looking to address this P2 issue: <https://github.com/LunarG/gfxreconstruct/issues/601>. I think the change will be incremental and systemic and we need to look out for opportunities to advance it as we work on other things. It would be good to get a revision of the description on there or close it if I shouldn't be keeping it in mind.

This is my last try to convince by words. I'll try not to bring up performance again without some profiling to show where we spend our time. For all I know we might be limited by the front end doing the decoding and all this string stuff is small potatoes in comparison. Looking for you to call it in your next reply :).


\* <https://github.com/LunarG/gfxreconstruct/issues/812>

Reply:
Never mind, without profiling, this is all moot, and now is not a good time to delve into that systematically. I'll make the change to use `std::strings` in this PR and make a note of this being one potential place to look for stripping out strings later, after some performance investigation.